### PR TITLE
Refactor Unit, Addon, Building, and more into interfaces

### PR DESCRIPTION
**Pull request by kdlin10**
Incorporates adakite's changes from dev branch. 

Refactored Unit from Class to an Interface. Also refactored Addon, Building, MobileUnit, GasMiningFacility, and PlayerUnit to interfaces. Moved implementations to abstract *Impl classes. Previously orphan interfaces such as TrainingFacility now inherit from new interfaces as appropriate. Existing classes now inherit from new *Impl abstract classes.

### CODE REVIEW

File: BWAPI4J/src/main/java/org/openbw/bwapi4j/unit/Addon.java [hunk 1]
```diff
@@ -1,36 +1,7 @@
-////////////////////////////////////////////////////////////////////////////////
-//
-//    Copyright (C) 2017-2018 OpenBW Team
```
Comment (adakitesystems) [inline BWAPI4J/src/main/java/org/openbw/bwapi4j/unit/Addon.java]: Copyright notices should be in all Java and C++ files regardless of their contents.

File: BWAPI4J/src/main/java/org/openbw/bwapi4j/unit/Building.java [hunk 2]
```diff
@@ -1,166 +1,34 @@
-////////////////////////////////////////////////////////////////////////////////
-//
-//    Copyright (C) 2017-2018 OpenBW Team
```
Comment (adakitesystems) [inline BWAPI4J/src/main/java/org/openbw/bwapi4j/unit/Building.java]: Copyright notices should be in all Java and C++ files regardless of their contents.

**Comment by adakitesystems** [comment]
Hi, @kdlin10! Thanks for the PR.

Overall I like the changes. One of the team members does not. The only minor nitpicks I have about this PR are copyright notices not being on all files and lack of `@Override` usage.

**Comment by kdlin10** [comment]
Hi @adakitesystems, I'd be happy to add in the copyright notices and `@Override`. Do I just upload updated ones?

I'm curious what the other team member doesn't like about the changes - in my mind, it shouldn't break any existing interfaces and extends functionality that helps avoid needless casting.

**Comment by Bytekeeper** [comment]
@kdlin10 Not liking is a bit too harsh. I'm not a huge fan of a mix of `Interfaces` and especially not of the `*Impl` "pattern" ;) It tends to lead to an (combinatorial) interface explosion.

For example, using your PR I now have `PlayerUnit` access without casting mostly. But what If I need a "more"? Ie. I want to build a `Spore`, so I need to find a `Creep Colony` first. If I do this in a generic way, I might only know it should be a `Building` and it should be `Morphable`. There's no `MorphableBuilding` interface, so I will still have to cast one or the other way.

**Comment by adakitesystems** [comment]
> I'd be happy to add in the copyright notices and @Override. Do I just upload updated ones?

It's OK. Don't worry about it this time. I'll add them. I want to apply some formatting to the entire Java codebase. The formatting will cause some problems which I'll fix myself.

**Comment by adakitesystems** [comment]
Merge conflicts needed to be resolved and the bridge needed to be fixed. Other than that, this PR has been successfully merged into the `develop` branch but was not automatically closed. Thanks for the PR, @muandrew!

**Comment by adakitesystems** [comment]
Merge conflicts needed to be resolved and the bridge needed to be fixed. Other than that, this PR has been successfully merged into the develop branch but was not automatically closed. Thanks for the PR, @kdlin10!

**Pull request closed without merging**